### PR TITLE
Allow overriding auto_update default for git materials in config repos (#2)

### PR DESCRIPTION
**Pull request by sandro-h**
It's now possible to set a "default auto_update" in the plugin config UI.
If a material doesn't have an explicit auto_update field in the gocd.yaml config, this
default value is used instead. This allows to override GoCD's default (`true`).
With this we can centrally control if git materials are in push or pull mode. Without it, we'd have to coordinate changing hundreds of yaml configs in projects.

The change is deliberately made to be minimally intrusive, i.e. without change a large number of files, to
simplify merging upstream changes (if upstream ever becomes active again). Hence the somewhat ugly DefaultOverrides
singleton. Still 19 changed files, but a bunch are tests.

I decided not to aim for a PR into upstream, see reasoning in ADNCD-1289.

### CODE REVIEW

File: src/main/java/cd/go/plugin/config/yaml/PluginSettings.java [hunk 1]
```diff
@@ -17,10 +22,20 @@
 
     static PluginSettings fromJson(String json) {
         Map<String, String> raw = JSONUtils.fromJSON(json);
-        return new PluginSettings(raw.get(PLUGIN_SETTINGS_FILE_PATTERN));
+        PluginSettings settings = new PluginSettings();
+        settings.filePattern = raw.get(PLUGIN_SETTINGS_FILE_PATTERN);
+        String defaultAutoUpdateVal = raw.get(PLUGIN_SETTINGS_DEFAULT_AUTO_UPDATE);
+        settings.defaultGitAutoUpdate = defaultAutoUpdateVal == null || defaultAutoUpdateVal.isBlank() ? null :
```
Comment (mcico) [inline src/main/java/cd/go/plugin/config/yaml/PluginSettings.java:28]: **question:** For me to understand, if no default value is set, the behavior of GoCD does not change, is that true? The plugin does **not** define a default value in the code directly?

**Review by mcico — CHANGES_REQUESTED** [review]

File: src/main/java/cd/go/plugin/config/yaml/PluginSettings.java (hunk 1 above)
Comment (sandro-h) [reply src/main/java/cd/go/plugin/config/yaml/PluginSettings.java:28]: Yes, if it is not set, the plugin doesn't even return an `autoUpdate` field to GoCD, which means GoCD uses its default.

**Comment by hosszubalazs** [comment]
I see your reasoning for the adn version and I agree makes sense here. Somewhere in docs drop a comment to highlight this requirement, later down the line we might be wondering we we needed the fork. :)

**Review by hosszubalazs — APPROVED** [review]

**Comment by sandro-h** [comment]
> I see your reasoning for the adn version and I agree makes sense here. Somewhere in docs drop a comment to highlight this requirement, later down the line we might be wondering we we needed the fork. :)

Yep, the fork and rationale is documented in Confluence under the GoCD Contribution Guide. But wouldn't hurt to add an explainer into the README on Github as well -> will do.

**Review by mcico — APPROVED** [review]

**Pull request merged**